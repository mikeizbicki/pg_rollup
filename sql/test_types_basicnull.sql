SET client_min_messages TO WARNING;
/*
 * this file tests the types in the basic test case, but puts more emphasis on null values;
 * there were some bugs where null values were not getting coalesced into zero,
 * and this file is designed to catch those edge cases;
 * it doesn't test more exotic types, but hopefully that's not necessary
 */
create or replace language plpython3u;
create extension if not exists pg_rollup;

create table nulltest (
    id serial primary key,
    name text,
    num int
);

insert into nulltest (name,num) values
    (NULL, NULL);

create materialized view nulltest_rollup1 as (
    select count(*),name
    from nulltest
    group by name
);

create materialized view nulltest_rollup2 as (
    select count(*),name,num
    from nulltest
    group by name,num
);

create materialized view nulltest_rollup3 as (
    select
        name,
        sum(num) as sum,
        count(*) as count_all,
        count(num) count_num,
        max(num),
        min(num)
    from nulltest
    group by name
);

create materialized view nulltest_rollup4 as (
    select 
        sum(num) as sum,
        count(*) as count_all,
        count(num),
        max(num),
        min(num)
    from nulltest
);

select assert_rollup('nulltest_rollup1');
select assert_rollup('nulltest_rollup2');
select assert_rollup('nulltest_rollup3');
select assert_rollup('nulltest_rollup4');

insert into nulltest (name,num) values
    (NULL, NULL);

select * from nulltest_rollup3_raw;
select * from nulltest_rollup3_groundtruth_raw;

select assert_rollup('nulltest_rollup1');
select assert_rollup('nulltest_rollup2');
select assert_rollup('nulltest_rollup3');
select assert_rollup('nulltest_rollup4');

insert into nulltest (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);

select * from nulltest_rollup3_raw;
select * from nulltest_rollup3_groundtruth_raw;

select assert_rollup('nulltest_rollup1');
select assert_rollup('nulltest_rollup2');
select assert_rollup('nulltest_rollup3');
select assert_rollup('nulltest_rollup4');

insert into nulltest (name,num) values
    (NULL, NULL);

select assert_rollup('nulltest_rollup1');
select assert_rollup('nulltest_rollup2');
select assert_rollup('nulltest_rollup3');
select assert_rollup('nulltest_rollup4');

insert into nulltest (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);

select assert_rollup('nulltest_rollup1');
select assert_rollup('nulltest_rollup2');
select assert_rollup('nulltest_rollup3');
select assert_rollup('nulltest_rollup4');


select * from nulltest_rollup3_raw;
select * from nulltest_rollup3_groundtruth_raw;

drop table nulltest cascade;
