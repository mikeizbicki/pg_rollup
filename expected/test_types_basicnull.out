SET client_min_messages TO WARNING;
/*
 * this file tests the types in the basic test case, but puts more emphasis on null values;
 * there were some bugs where null values were not getting coalesced into zero,
 * and this file is designed to catch those edge cases;
 * it doesn't test more exotic types, but hopefully that's not necessary
 */
create table nulltest (
    id serial primary key,
    name text,
    num int
);
insert into nulltest (name,num) values
    (NULL, NULL);
create materialized view nulltest_rollup1 as (
    select count(*),name
    from nulltest
    group by name
);
create materialized view nulltest_rollup2 as (
    select count(*),name,num
    from nulltest
    group by name,num
);
create materialized view nulltest_rollup3 as (
    select
        name,
        sum(num) as sum,
        count(*) as count_all,
        count(num) count_num,
        max(num),
        min(num)
    from nulltest
    group by name
);
create materialized view nulltest_rollup4 as (
    select 
        sum(num) as sum,
        count(*) as count_all,
        count(num),
        max(num),
        min(num)
    from nulltest
);
select assert_rollup('nulltest_rollup1');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup2');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup3');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup4');
 assert_rollup 
---------------
 
(1 row)

insert into nulltest (name,num) values
    (NULL, NULL);
select * from nulltest_rollup3_raw;
 count(*) | count(nulltest.num) | max(nulltest.num) | min(nulltest.num) | sum(nulltest.num) | nulltest.name 
----------+---------------------+-------------------+-------------------+-------------------+---------------
        2 |                   0 |                   |                   |                 0 | 
(1 row)

select * from nulltest_rollup3_groundtruth_raw;
 count(*) | count(nulltest.num) | max(nulltest.num) | min(nulltest.num) | sum(nulltest.num) | nulltest.name 
----------+---------------------+-------------------+-------------------+-------------------+---------------
        2 |                   0 |                   |                   |                 0 | 
(1 row)

select assert_rollup('nulltest_rollup1');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup2');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup3');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup4');
 assert_rollup 
---------------
 
(1 row)

insert into nulltest (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
select * from nulltest_rollup3_raw;
 count(*) | count(nulltest.num) | max(nulltest.num) | min(nulltest.num) | sum(nulltest.num) | nulltest.name 
----------+---------------------+-------------------+-------------------+-------------------+---------------
       13 |                   7 |                 9 |                 1 |                22 | 
        5 |                   5 |                 5 |                 1 |                15 | alice
        1 |                   1 |                 5 |                 5 |                 5 | elliot
        1 |                   1 |                 4 |                 4 |                 4 | dave
        3 |                   3 |                 5 |                 5 |                15 | bill
        5 |                   4 |                 3 |                 1 |                 6 | charlie
(6 rows)

select * from nulltest_rollup3_groundtruth_raw;
 count(*) | count(nulltest.num) | max(nulltest.num) | min(nulltest.num) | sum(nulltest.num) | nulltest.name 
----------+---------------------+-------------------+-------------------+-------------------+---------------
       13 |                   7 |                 9 |                 1 |                22 | 
        5 |                   4 |                 3 |                 1 |                 6 | charlie
        1 |                   1 |                 4 |                 4 |                 4 | dave
        3 |                   3 |                 5 |                 5 |                15 | bill
        1 |                   1 |                 5 |                 5 |                 5 | elliot
        5 |                   5 |                 5 |                 1 |                15 | alice
(6 rows)

select assert_rollup('nulltest_rollup1');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup2');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup3');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup4');
 assert_rollup 
---------------
 
(1 row)

insert into nulltest (name,num) values
    (NULL, NULL);
select assert_rollup('nulltest_rollup1');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup2');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup3');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup4');
 assert_rollup 
---------------
 
(1 row)

insert into nulltest (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
select assert_rollup('nulltest_rollup1');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup2');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup3');
 assert_rollup 
---------------
 
(1 row)

select assert_rollup('nulltest_rollup4');
 assert_rollup 
---------------
 
(1 row)

select * from nulltest_rollup3_raw;
 count(*) | count(nulltest.num) | max(nulltest.num) | min(nulltest.num) | sum(nulltest.num) | nulltest.name 
----------+---------------------+-------------------+-------------------+-------------------+---------------
       25 |                  14 |                 9 |                 1 |                44 | 
       10 |                  10 |                 5 |                 1 |                30 | alice
        2 |                   2 |                 5 |                 5 |                10 | elliot
        2 |                   2 |                 4 |                 4 |                 8 | dave
        6 |                   6 |                 5 |                 5 |                30 | bill
       10 |                   8 |                 3 |                 1 |                12 | charlie
(6 rows)

select * from nulltest_rollup3_groundtruth_raw;
 count(*) | count(nulltest.num) | max(nulltest.num) | min(nulltest.num) | sum(nulltest.num) | nulltest.name 
----------+---------------------+-------------------+-------------------+-------------------+---------------
       25 |                  14 |                 9 |                 1 |                44 | 
       10 |                   8 |                 3 |                 1 |                12 | charlie
        2 |                   2 |                 4 |                 4 |                 8 | dave
        6 |                   6 |                 5 |                 5 |                30 | bill
        2 |                   2 |                 5 |                 5 |                10 | elliot
       10 |                  10 |                 5 |                 1 |                30 | alice
(6 rows)

drop table nulltest cascade;
